generator/intermediate_data.cpp

```diff
@@ -266,5 +263,94 @@ void MapFilePointStorageWriter::AddPoint(uint64_t id, double lat, double lon)
 
   ++m_numProcessedPoints;
 }
+
+// IntermediateDataReader
+IntermediateDataReader::IntermediateDataReader(shared_ptr<IPointStorageReader> nodes,
+                                               feature::GenerateInfo & info) :
+  m_nodes(nodes),
+  m_ways(info.GetIntermediateFileName(WAYS_FILE), info.m_preloadCache),
+  m_relations(info.GetIntermediateFileName(RELATIONS_FILE), info.m_preloadCache),
+  m_nodeToRelations(info.GetIntermediateFileName(NODES_FILE, ID2REL_EXT)),
+  m_wayToRelations(info.GetIntermediateFileName(WAYS_FILE, ID2REL_EXT))
+{
+}
+
+void IntermediateDataReader::LoadIndex()
+{
+  m_ways.LoadOffsets();
+  m_relations.LoadOffsets();
+
+  m_nodeToRelations.ReadAll();
+  m_wayToRelations.ReadAll();
+}
+
+
+// IntermediateDataWriter
+IntermediateDataWriter::IntermediateDataWriter(std::shared_ptr<IPointStorageWriter> nodes,
+                                               feature::GenerateInfo & info):
+  m_nodes(nodes),
+  m_ways(info.GetIntermediateFileName(WAYS_FILE), info.m_preloadCache),
+  m_relations(info.GetIntermediateFileName(RELATIONS_FILE), info.m_preloadCache),
+  m_nodeToRelations(info.GetIntermediateFileName(NODES_FILE, ID2REL_EXT)),
+  m_wayToRelations(info.GetIntermediateFileName(WAYS_FILE, ID2REL_EXT))
+{
+}
+
+void IntermediateDataWriter::AddRelation(Key id, RelationElement const & e)
+{
+  string const & relationType = e.GetType();
+  if (!(relationType == "multipolygon" || relationType == "route" || relationType == "boundary" ||
```

Comment:
Maybe create `std::set` somewhere outside and use it?

Reply:
++